strawberry_django/filters.py

```diff
@@ -1,147 +1,165 @@
+        if filter_method:
+            filter_methods.append(filter_method)
+            continue
+
+        if django_model:
+            if field_name not in django_model._meta._forward_fields_map:
+                continue
+
+        if field_name in lookup_name_conversion_map:
+            field_name = lookup_name_conversion_map[field_name]
+        if utils.is_strawberry_type(field_value):
+            subfield_filter_kwargs, subfield_filter_methods = \
+                build_filter_kwargs(field_value)
+            for subfield_name, subfield_value in subfield_filter_kwargs.items():
+                filter_kwargs[f'{field_name}__{subfield_name}'] = subfield_value
+            filter_methods.extend(subfield_filter_methods)
+        else:
+            filter_kwargs[field_name] = field_value
+
+    return filter_kwargs, filter_methods
+
+
+def apply(filters, queryset, pk=UNSET):
+    if not is_unset(pk):
+        queryset = queryset.filter(pk=pk)
+    if is_unset(filters) or filters is None:
         return queryset
 
-    data = {}
-    for field_name in filter_instance.filterset_class.get_fields():
-        value = getattr(filter_instance, field_name, None)
-        if not is_unset(value):
-            data[field_name] = value
-
-    filterset = filter_instance.filterset_class(
-        data=data,
-        queryset=queryset,
-    )
-
-    if not filterset.is_valid():
-        raise InvalidFilterError(filterset.errors)
-
-    return filterset.qs
-
-
-@assert_django_filters_installed
-def filter(filterset_class: Type[django_filters.FilterSet], name=None):
-    if not isinstance(filterset_class, django_filters.filterset.FilterSetMetaclass):
-        raise TypeError(
-            "strawberry_django.filter expects a class that inherits django_filters.FilterSet, received %s",
-            type(filterset_class),
-        )
-
-    filters = filterset_class.get_filters()
-    name = name or filterset_class.__name__
-    cls = type(name, (), {"__annotations__": {}, "filterset_class": filterset_class})
-
-    for field_name in filterset_class.get_filters().keys():
-        filter_field = filters[field_name]
-        field_type = get_field_type(type(filter_field))
-        cls.__annotations__[field_name] = field_type
-        setattr(cls, field_name, UNSET)
-
-    return strawberry.input(cls)
+    if hasattr(filters, 'filterset_class'):
+        utils.deprecated("support for 'django-filters' is deprecated and"
+            " will removed in v0.3", stacklevel=2)
+        from .legacy.filters import apply as filters_apply
+        return filters_apply(filters, queryset)
+
+    filter_method = getattr(filters, 'filter', None)
+    if filter_method:
+        return filter_method(queryset)
+
+    filter_kwargs, filter_methods = build_filter_kwargs(filters)
+    queryset = queryset.filter(**filter_kwargs)
+    for filter_method in filter_methods:
+        queryset = filter_method(queryset=queryset)
+    return queryset
+
+
+class StrawberryDjangoFieldFilters:
```

Comment:
Maybe we should have a base type for all these mixins

Reply:
We have `StrawberryDjangoFieldBase` class in https://github.com/strawberry-graphql/strawberry-graphql-django/blob/feature/class-api/strawberry_django/fields/field.py#L12-L18.

Probably these should be inherited from that because they are using functions from there.